src/events/mod.rs

```diff
@@ -404,7 +456,41 @@ impl<'a> BytesDecl<'a> {
         }
     }
 
-    /// Gets xml encoding, including quotes (' or ")
+    /// Gets xml encoding, excluding quotes (' or ")
+    ///
+    /// Although according to the [grammar] encoding must appear before `"standalone"`
+    /// and after `"version"`, this method does not check that. The first occurrence
+    /// of the attribute will be returned even if there are several. Also, method does
+    /// not restrict symbols that can forming the encoding, so the returned encoding
+    /// name may not correspond to the grammar.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use std::borrow::Cow;
+    /// use quick_xml::Error;
+    /// use quick_xml::events::{BytesDecl, BytesStart};
+    ///
+    /// // <?xml version='1.1'?>
+    /// let decl = BytesDecl::from_start(BytesStart::borrowed(b" version='1.1'", 0));
+    /// assert!(decl.encoding().is_none());
+    ///
+    /// // <?xml encoding='utf-8'?>
+    /// let decl = BytesDecl::from_start(BytesStart::borrowed(b" encoding='utf-8'", 0));
+    /// match decl.encoding() {
+    ///     Some(Ok(Cow::Borrowed(encoding))) => assert_eq!(encoding, b"utf-8"),
+    ///     _ => assert!(false),
+    /// }
+    ///
+    /// // <?xml encoding='something_WRONG' encoding='utf-8'?>
+    /// let decl = BytesDecl::from_start(BytesStart::borrowed(b" encoding='something_WRONG' encoding='utf-8'", 0));
+    /// match decl.encoding() {
+    ///     Some(Ok(Cow::Borrowed(encoding))) => assert_eq!(encoding, b"something_WRONG"),
+    ///     _ => assert!(false),
+    /// }
+    /// ```
+    ///
+    /// [grammar]: https://www.w3.org/TR/xml11/#NT-XMLDecl
     pub fn encoding(&self) -> Option<Result<Cow<[u8]>>> {
         for a in self.element.attributes() {
```

Comment:
For these two methods you can probably simplify the implementation by using [`BytesStart::try_get_attribute()`](https://github.com/tafia/quick-xml/blob/521d72a28449eaa0fcb6a2cbeaead72921d92b81/src/events/mod.rs#L352-L364=) instead.

And as a bonus it will get a speedup from my PR whenever it gets merged. https://github.com/tafia/quick-xml/pull/348

You can't do that for `::version()` above because of the requirement to check that the version attribute comes first.  But you can still do some optimizatoin by setting `.with_checks(false)` manually.

Reply:
Good catch! Then I think you should just include that change into your PR, because it is already about optimisations

Reply:
Fair point, I'll do that

Reply:
I think, it will be better. The `Result<Option<...>>` seems more idiomatic and opens a door for an internal changes that could require to return an error even if no attributes defined, if that changes will be necessary sometime

Reply:
Separate PR: https://github.com/tafia/quick-xml/pull/378/